File: android/app/src/main/java/com/simpleshift/scheduler/domain/leave_optimizer.kt
package com.simpleshift.scheduler.domain

import com.simpleshift.scheduler.domain.model.LeaveStrategy
import com.simpleshift.scheduler.domain.model.ShiftCycleConfig
import com.simpleshift.scheduler.domain.model.ShiftType
import com.simpleshift.scheduler.domain.model.Team
import java.time.LocalDate
import java.time.temporal.ChronoUnit

// ── Internal day status used during analysis ──

internal data class DayStatus(
    val date: LocalDate,
    val isRest: Boolean,
    val isHoliday: Boolean,
    val isWeekend: Boolean,
    val isAdjustedWorkDay: Boolean,
    val holidayName: String?
)

// ── Public API ──

/**
 * Finds the best leave strategies for the next [daysToAnalyze] days.
 *
 * Algorithm: Gap-merging + extension.
 * 1. Build daily status for the analysis window (shift schedule + holidays + weekends).
 * 2. Scan for work gaps (consecutive work days) and bridge them with leave.
 * 3. Also extend rest blocks by taking leave on adjacent work days.
 * 4. Score, deduplicate, and rank.
 *
 * @param today Start date for analysis (inclusive).
 * @param daysToAnalyze Number of days to look ahead (default 365).
 * @param teamPhaseOffset Team phase offset for shift calculation.
 * @param customCycle Custom shift cycle, uses default if null.
 * @param referenceDate Reference date for the shift cycle.
 * @param holidays China holiday data map (date -> info).
 * @param maxLeaveDays Maximum leave days to consider per strategy.
 */
fun findBestLeavePlans(
    today: LocalDate = LocalDate.now(),
    daysToAnalyze: Int = 365,
    teamPhaseOffset: Int = 0,
    customCycle: List<ShiftType>? = null,
    referenceDate: LocalDate = ShiftCycleConfig.REFERENCE_DATE,
    holidays: Map<LocalDate, HolidayInfo> = getChinaHolidays(),
    maxLeaveDays: Int = 5
): List<LeaveStrategy> {
    if (daysToAnalyze < 1 || maxLeaveDays < 1) return emptyList()

    val status = buildDailyStatus(
        startDate = today,
        days = daysToAnalyze,
        teamPhaseOffset = teamPhaseOffset,
        customCycle = customCycle,
        referenceDate = referenceDate,
        holidays = holidays
    )

    // Precompute consecutive rest before/after each index
    val n = status.size
    val restBefore = IntArray(n)
    val restAfter = IntArray(n)

    for (i in 1 until n) {
        restBefore[i] = if (status[i - 1].isRest) restBefore[i - 1] + 1 else 0
    }
    for (i in n - 2 downTo 0) {
        restAfter[i] = if (status[i + 1].isRest) restAfter[i + 1] + 1 else 0
    }

    val strategies = mutableListOf<LeaveStrategy>()

    for (leaveDays in 1..maxLeaveDays) {
        for (startIdx in 0..(n - leaveDays)) {
            // All leave days must be work days (not rest)
            if ((0 until leaveDays).any { status[startIdx + it].isRest }) continue

            // All leave days must be within the analysis window and not in the past
            // (already guaranteed by startIdx range)

            val leftRest = restBefore[startIdx]
            val rightRest = restAfter[startIdx + leaveDays - 1]

            // The break: left rest + leave days + right rest
            val totalBreak = leftRest + leaveDays + rightRest
            if (totalBreak <= leaveDays) continue // no benefit

            val gapStart = startIdx - leftRest
            val gapEnd = startIdx + leaveDays - 1 + rightRest

            val breakStartDate = status[gapStart].date
            val breakEndDate = status[gapEnd].date

            // Calculate family overlap within the break
            var holidayOverlap = 0
            var weekendOverlap = 0
            val holidayNames = mutableSetOf<String>()

            for (idx in gapStart..gapEnd) {
                val ds = status[idx]
                if (ds.isHoliday) {
                    holidayOverlap++
                    ds.holidayName?.let { holidayNames.add(it) }
                }
                if (ds.isWeekend && !ds.isAdjustedWorkDay) {
                    weekendOverlap++
                }
            }

            val leaveDateList = (startIdx until startIdx + leaveDays).map { status[it].date }

            val efficiency = totalBreak.toFloat() / leaveDays

            strategies.add(
                LeaveStrategy(
                    leaveDays = leaveDays,
                    totalBreakDays = totalBreak,
                    leaveDates = leaveDateList,
                    breakStart = breakStartDate,
                    breakEnd = breakEndDate,
                    holidayOverlap = holidayOverlap,
                    weekendOverlap = weekendOverlap,
                    overlappingHolidayNames = holidayNames.toList(),
                    efficiency = efficiency,
                    score = 0f // computed later
                )
            )
        }
    }

    // Deduplicate: same (breakStart, breakEnd) → keep the one with fewest leave days
    val deduped = strategies
        .groupBy { Pair(it.breakStart, it.breakEnd) }
        .mapValues { (_, list) -> list.minByOrNull { it.leaveDays }!! }
        .values
        .toList()

    if (deduped.isEmpty()) return emptyList()

    // Compute scores
    val maxEfficiency = deduped.maxOf { it.efficiency }
    val maxBreak = deduped.maxOf { it.totalBreakDays }
    val maxFamilyBonus = deduped.maxOf { it.holidayOverlap * 2 + it.weekendOverlap }.coerceAtLeast(1)

    val scored = deduped.map { strategy ->
        val effScore = if (maxEfficiency > 0f) strategy.efficiency / maxEfficiency else 0f
        val lenScore = if (maxBreak > 0) strategy.totalBreakDays.toFloat() / maxBreak else 0f
        val familyBonus = strategy.holidayOverlap * 2 + strategy.weekendOverlap
        val famScore = familyBonus.toFloat() / maxFamilyBonus

        val score = 0.50f * effScore + 0.25f * lenScore + 0.25f * famScore
        strategy.copy(score = score)
    }

    return scored.sortedByDescending { it.score }
}

// ── Internal helpers ──

internal fun buildDailyStatus(
    startDate: LocalDate,
    days: Int,
    teamPhaseOffset: Int,
    customCycle: List<ShiftType>?,
    referenceDate: LocalDate,
    holidays: Map<LocalDate, HolidayInfo>
): List<DayStatus> {
    return (0 until days).map { offset ->
        val date = startDate.plusDays(offset.toLong())
        val shiftType = getShiftTypeForDate(date, teamPhaseOffset, customCycle, referenceDate)
        val isRest = shiftType == ShiftType.REST || shiftType == ShiftType.STUDY
        val holidayInfo = holidays[date]
        val isHoliday = holidayInfo?.isHoliday == true
        val isAdjustedWorkDay = holidayInfo != null && !holidayInfo.isHoliday

        DayStatus(
            date = date,
            isRest = isRest,
            isHoliday = isHoliday,
            isWeekend = isWeekend(date),
            isAdjustedWorkDay = isAdjustedWorkDay,
            holidayName = if (isHoliday) holidayInfo?.name else null
        )
    }
}

/**
 * Returns the team phase offset for a given team ID and cycle.
 */
fun teamPhaseOffsetFor(teamId: Int, customCycle: List<ShiftType>? = null): Int {
    val totalDays = customCycle?.size ?: ShiftCycleConfig.CYCLE_LENGTH
    return (teamId - 1) * (totalDays / Team.TOTAL_TEAMS)
}

File: flutter/lib/domain/algorithms/leave_optimizer.dart
import '../models/leave_strategy.dart';
import '../models/shift_cycle_config.dart';
import '../models/shift_type.dart';
import 'shift_calculator.dart';
import 'holiday_data.dart';

// ── Internal day status ──

class DayStatus {
  final DateTime date;
  final bool isRest;
  final bool isHoliday;
  final bool isWeekend;
  final bool isAdjustedWorkDay;
  final String? holidayName;

  const DayStatus({
    required this.date,
    required this.isRest,
    required this.isHoliday,
    required this.isWeekend,
    required this.isAdjustedWorkDay,
    this.holidayName,
  });

  bool get isOff =>
      isRest ||
      (isHoliday && !isAdjustedWorkDay) ||
      (isWeekend && !isAdjustedWorkDay);
}

// ── Internal helpers ──

List<DayStatus> buildDailyStatus({
  required DateTime startDate,
  required int days,
  int teamPhaseOffset = 0,
  List<ShiftType>? customCycle,
  DateTime? referenceDate,
  required Map<DateTime, HolidayInfo> holidays,
}) {
  return List.generate(days, (offset) {
    final date = startDate.add(Duration(days: offset));
    final shiftType = getShiftTypeForDate(
      date,
      teamPhaseOffset: teamPhaseOffset,
      customCycle: customCycle,
      referenceDate: referenceDate,
    );
    final isRest =
        shiftType == ShiftType.REST || shiftType == ShiftType.STUDY;
    final holidayInfo = holidays[date];
    final isHoliday = holidayInfo?.isHoliday == true;
    final isAdjustedWorkDay =
        holidayInfo != null && !holidayInfo.isHoliday;

    return DayStatus(
      date: date,
      isRest: isRest,
      isHoliday: isHoliday,
      isWeekend: isWeekend(date),
      isAdjustedWorkDay: isAdjustedWorkDay,
      holidayName: isHoliday ? holidayInfo?.name : null,
    );
  });
}

// ── Public API ──

/// 间隙桥接法：查找最佳请假方案 — 对应 Android 版 findBestLeavePlans()
List<LeaveStrategy> findBestLeavePlans({
  DateTime? today,
  int daysToAnalyze = 365,
  int teamPhaseOffset = 0,
  List<ShiftType>? customCycle,
  DateTime? referenceDate,
  Map<DateTime, HolidayInfo>? holidays,
  int maxLeaveDays = 5,
}) {
  if (daysToAnalyze < 1 || maxLeaveDays < 1) return [];

  final raw = today ?? DateTime.now();
  // Normalize to midnight: DateTime(2026, 6, 19, 0, 0, 0) matches holiday map keys
  final t = DateTime(raw.year, raw.month, raw.day);
  final ref = referenceDate ?? ShiftCycleConfig.referenceDate;
  final hols = holidays ?? getChinaHolidays();

  final status = buildDailyStatus(
    startDate: t,
    days: daysToAnalyze,
    teamPhaseOffset: teamPhaseOffset,
    customCycle: customCycle,
    referenceDate: ref,
    holidays: hols,
  );

  final n = status.length;

  // Precompute consecutive rest before/after each index
  final restBefore = List.filled(n, 0);
  final restAfter = List.filled(n, 0);

  for (int i = 1; i < n; i++) {
    restBefore[i] = status[i - 1].isOff ? restBefore[i - 1] + 1 : 0;
  }
  for (int i = n - 2; i >= 0; i--) {
    restAfter[i] = status[i + 1].isOff ? restAfter[i + 1] + 1 : 0;
  }

  final strategies = <LeaveStrategy>[];

  // Start from 2: single-day leave is trivial (user doesn't need the app for that).
  // Unless maxLeaveDays == 1 (user explicitly filtered to 1).
  final minLeaveDays = maxLeaveDays == 1 ? 1 : 2;
  for (int leaveDays = minLeaveDays; leaveDays <= maxLeaveDays; leaveDays++) {
    for (int startIdx = 0; startIdx <= n - leaveDays; startIdx++) {
      // All leave days must be work days (only exclude shift rest/study, not weekends)
      // Weekends and holidays are naturally off but can be bridged without leave
      var anyRest = false;
      for (int j = 0; j < leaveDays; j++) {
        if (status[startIdx + j].isRest) {
          anyRest = true;
          break;
        }
      }
      if (anyRest) continue;

      final leftRest = restBefore[startIdx];
      final rightRest = restAfter[startIdx + leaveDays - 1];

      final totalBreak = leftRest + leaveDays + rightRest;
      if (totalBreak <= leaveDays) continue;

      final gapStart = startIdx - leftRest;
      final gapEnd = startIdx + leaveDays - 1 + rightRest;

      final breakStartDate = status[gapStart].date;
      final breakEndDate = status[gapEnd].date;

      // Calculate family overlap
      int holidayOverlap = 0;
      int weekendOverlap = 0;
      final holidayNames = <String>{};

      for (int idx = gapStart; idx <= gapEnd; idx++) {
        final ds = status[idx];
        if (ds.isHoliday) {
          holidayOverlap++;
          if (ds.holidayName != null) holidayNames.add(ds.holidayName!);
        }
        if (ds.isWeekend && !ds.isAdjustedWorkDay) {
          weekendOverlap++;
        }
      }

      final leaveDateList = List.generate(
          leaveDays, (j) => status[startIdx + j].date);

      final efficiency = totalBreak / leaveDays;

      strategies.add(LeaveStrategy(
        leaveDays: leaveDays,
        totalBreakDays: totalBreak,
        leaveDates: leaveDateList,
        breakStart: breakStartDate,
        breakEnd: breakEndDate,
        holidayOverlap: holidayOverlap,
        weekendOverlap: weekendOverlap,
        overlappingHolidayNames: holidayNames.toList(),
        efficiency: efficiency,
        score: 0,
      ));
    }
  }

  // Deduplicate: same (breakStart, breakEnd) → keep fewest leave days
  final deduped = <String, LeaveStrategy>{};
  for (final s in strategies) {
    final key = '${s.breakStart}_${s.breakEnd}';
    if (!deduped.containsKey(key) || deduped[key]!.leaveDays > s.leaveDays) {
      deduped[key] = s;
    }
  }
  if (deduped.isEmpty) return [];

  final dedupedList = deduped.values.toList();

  // Compute scores
  final maxEfficiency = dedupedList
      .map((s) => s.efficiency)
      .reduce((a, b) => a > b ? a : b);
  final maxBreak = dedupedList
      .map((s) => s.totalBreakDays)
      .reduce((a, b) => a > b ? a : b);
  final maxFamilyBonus = dedupedList
      .map((s) => s.holidayOverlap * 2 + s.weekendOverlap)
      .reduce((a, b) => a > b ? a : b)
      .clamp(1, 999);

  final scored = dedupedList.map((strategy) {
    final effScore = maxEfficiency > 0 ? strategy.efficiency / maxEfficiency : 0.0;
    final lenScore = maxBreak > 0 ? strategy.totalBreakDays / maxBreak : 0.0;
    final familyBonus = strategy.holidayOverlap * 2 + strategy.weekendOverlap;
    final famScore = familyBonus / maxFamilyBonus;
    final score = 0.50 * effScore + 0.25 * lenScore + 0.25 * famScore;
    return LeaveStrategy(
      leaveDays: strategy.leaveDays,
      totalBreakDays: strategy.totalBreakDays,
      leaveDates: strategy.leaveDates,
      breakStart: strategy.breakStart,
      breakEnd: strategy.breakEnd,
      holidayOverlap: strategy.holidayOverlap,
      weekendOverlap: strategy.weekendOverlap,
      overlappingHolidayNames: strategy.overlappingHolidayNames,
      efficiency: strategy.efficiency,
      score: score,
    );
  }).toList();

  scored.sort((a, b) => b.score.compareTo(a.score));
  return scored;
}
